From 7d06934b68ef6e14e113df95b78b38d619cb7295 Mon Sep 17 00:00:00 2001
From: Sathyanarayanan Kuppuswamy <sathyanarayanan.kuppuswamy@intel.com>
Date: Fri, 25 Oct 2013 13:52:41 -0700
Subject: [PATCH 02/11] AndroidIA: power_hal: Add sysfs based power_hal suspend
 support

This patch implements the sysfs based PowerHAL interface that
deprecates early suspend support that has been removed.

Added a new sysfs group for power_hal suspend under
/sys/power/ directory. Every device that needs power_hal
suspend support needs to call following routines
to register/unregister power_hal suspend support.

register_power_hal_suspend_device():
Registers power_hal suspend support for the given device. Also,
Calling this routine will create a device specific symlink
under /sys/power/power_hal directory.

unregister_power_hal_suspend_device():
Unregisters power_hal suspend support for the given device. Also,
Calling this routine will remove a device specific symlink
under /sys/power/power_hal_suspend directory.

`ls /sys/power/power_hal/` from the userspace will give
us the list of devices that supports power_hal suspend functionality.

Change-Id: I7f15c42fc8baa9bf7db8be6b77b6622ca5d3a648
Signed-off-by: Sathyanarayanan Kuppuswamy <sathyanarayanan.kuppuswamy@intel.com>
---
 include/linux/power_hal_sysfs.h | 18 ++++++++++++++++++
 kernel/power/main.c             | 28 ++++++++++++++++++++++++++++
 2 files changed, 46 insertions(+)
 create mode 100644 include/linux/power_hal_sysfs.h

diff --git a/include/linux/power_hal_sysfs.h b/include/linux/power_hal_sysfs.h
new file mode 100644
index 0000000..ecb8d85
--- /dev/null
+++ b/include/linux/power_hal_sysfs.h
@@ -0,0 +1,18 @@
+/*
+ * power_hal_sysfs.h: Early suspend sysfs header file
+ *
+ * (C) Copyright 2013 Intel Corporation
+ * Author: Sathyanarayanan KN(sathyanarayanan.kuppuswamy@intel.com)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; version 2
+ * of the License.
+ */
+
+#ifndef _LINUX_POWER_HAL_SUSPEND_SYSFS_H
+#define _LINUX_POWER_HAL_SUSPEND_SYSFS_H
+
+int register_power_HAL_suspend_device(struct device *dev);
+void unregister_power_HAL_suspend_device(struct device *dev);
+#endif
diff --git a/kernel/power/main.c b/kernel/power/main.c
index 3a2ca90..d053b3b 100644
--- a/kernel/power/main.c
+++ b/kernel/power/main.c
@@ -15,6 +15,7 @@
 #include <linux/workqueue.h>
 #include <linux/debugfs.h>
 #include <linux/seq_file.h>
+#include <linux/power_hal_sysfs.h>
 
 #include "power.h"
 
@@ -421,6 +422,7 @@ static inline void pm_print_times_init(void) {}
 #endif /* CONFIG_PM_SLEEP_DEBUG */
 
 struct kobject *power_kobj;
+struct kobject *power_hal_kobj;
 
 /**
  * state - control system sleep states.
@@ -773,6 +775,28 @@ static int __init pm_start_workqueue(void)
 	return pm_wq ? 0 : -ENOMEM;
 }
 
+int register_power_hal_suspend_device(struct device *dev)
+{
+	int ret;
+	if (!power_hal_kobj || !dev)
+		return -ENODEV;
+
+	ret = sysfs_create_link(power_hal_kobj, &dev->kobj,
+			dev_name(dev));
+
+	kobject_uevent(&dev->kobj, KOBJ_CHANGE);
+
+	return ret;
+}
+EXPORT_SYMBOL(register_power_hal_suspend_device);
+
+void unregister_power_hal_suspend_device(struct device *dev)
+{
+	sysfs_delete_link(power_hal_kobj, &dev->kobj, dev_name(dev));
+}
+EXPORT_SYMBOL(unregister_power_hal_suspend_device);
+
+
 static int __init pm_init(void)
 {
 	int error = pm_start_workqueue();
@@ -782,8 +806,12 @@ static int __init pm_init(void)
 	hibernate_reserved_size_init();
 	pm_states_init();
 	power_kobj = kobject_create_and_add("power", NULL);
+	power_hal_kobj = kobject_create_and_add("power_HAL_suspend",
+					power_kobj);
 	if (!power_kobj)
 		return -ENOMEM;
+	if (!power_hal_kobj)
+		return -ENOMEM;
 	error = sysfs_create_group(power_kobj, &attr_group);
 	if (error)
 		return error;
-- 
1.9.1

